File: src/main/java/com/github/hirsivaja/ip/ipv4/Ipv4Header.java
package com.github.hirsivaja.ip.ipv4;

import com.github.hirsivaja.ip.EcnCodePoint;
import com.github.hirsivaja.ip.IpHeader;
import com.github.hirsivaja.ip.IpProtocol;
import com.github.hirsivaja.ip.IpUtils;
import com.github.hirsivaja.ip.ipv4.option.IpOption;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;

public record Ipv4Header(
        byte dscp,
        EcnCodePoint ecn,
        short len,
        short identification,
        Ipv4Flags flags,
        short fragmentOffset,
        byte ttl,
        IpProtocol protocol,
        Ipv4Address srcIp,
        Ipv4Address dstIp,
        List<IpOption> options) implements IpHeader {
    public static final byte VERSION = (byte) 0x04;
    public static final int HEADER_LEN = 20;
    public static final int PSEUDO_HEADER_LEN = 12;
    public static final int VERSION_SHIFT = 4;
    private static final int DSCP_SHIFT = 2;
    private static final int FLAGS_SHIFT = 13;

    @SuppressWarnings("squid:S00107")
    public Ipv4Header(byte dscp, EcnCodePoint ecn, short len, short identification, Ipv4Flags flags, short fragmentOffset, byte ttl,
                      IpProtocol protocol, Ipv4Address srcIp, Ipv4Address dstIp) {
        this(dscp, ecn, len, identification, flags, fragmentOffset, ttl, protocol, srcIp, dstIp, List.of());
    }

    @Override
    public void encode(ByteBuffer out) {
        out.mark();
        byte versionIhl = (byte) (VERSION << VERSION_SHIFT | ((optionsLength() / 4) + 5));
        out.put(versionIhl);
        byte dscpEcn = (byte) (ecn.type() & 0xFF | (dscp << DSCP_SHIFT));
        out.put(dscpEcn);
        out.putShort(len);
        out.putShort(identification);
        short flagsFragmentOffset = (short) (fragmentOffset | (flags.toByte() << FLAGS_SHIFT));
        out.putShort(flagsFragmentOffset);
        out.put(ttl);
        out.put(protocol.type());
        int checksumPosition = out.position();
        out.putShort((short) 0);
        srcIp.encode(out);
        dstIp.encode(out);
        options.forEach(option -> option.encode(out));
        int position = out.position();
        byte[] headerBytes = new byte[HEADER_LEN];
        out.reset().get(headerBytes);
        short checksum = IpUtils.calculateInternetChecksum(headerBytes);
        out.putShort(checksumPosition, checksum);
        out.position(position);
    }

    @Override
    public byte[] generatePseudoHeader() {
        ByteBuffer out = ByteBuffer.allocate(PSEUDO_HEADER_LEN);
        srcIp.encode(out);
        dstIp.encode(out);
        out.put((byte) 0);
        out.put(protocol.type());
        out.putShort((short) (len - HEADER_LEN - optionsLength()));
        byte[] outBytes = new byte[PSEUDO_HEADER_LEN];
        out.rewind().get(outBytes);
        return outBytes;
    }

    /**
     * Header + options length (no payload)
     */
    @Override
    public int length() {
        return HEADER_LEN + optionsLength();
    }

    /**
     * Payload + options length (no header)
     */
    public int dataLength() {
        return Short.toUnsignedInt(len) - Ipv4Header.HEADER_LEN;
    }

    /**
     * Payload + options + header length
     */
    public int totalLength() {
        return Short.toUnsignedInt(len);
    }

    /**
     * Payload length (no header or options)
     */
    public int payloadLength() {
        return totalLength() - Ipv4Header.HEADER_LEN - optionsLength();
    }

    private int optionsLength() {
        return options.stream().mapToInt(IpOption::length).sum();
    }

    @Override
    public int pseudoHeaderLength() {
        return PSEUDO_HEADER_LEN;
    }

    public static Ipv4Header decode(ByteBuffer in) {
        return decode(in, true);
    }

    public static Ipv4Header decode(ByteBuffer in, boolean ensureChecksum) {
        in.mark();
        byte version = in.get();
        byte ihl = (byte) (version & 0x0F);
        version = (byte) (version >>> VERSION_SHIFT);
        if(version != VERSION){
            throw new IllegalArgumentException("Unexpected version for IPv4 header! " + version);
        }
        byte dscp = in.get();
        EcnCodePoint ecn = EcnCodePoint.fromType((byte) (dscp & 0b11));
        dscp = (byte) (dscp >>> DSCP_SHIFT);
        short len = in.getShort();
        short identification = in.getShort();
        short fragmentOffset = in.getShort();
        Ipv4Flags flags = Ipv4Flags.decode((byte) (fragmentOffset >>> FLAGS_SHIFT));
        fragmentOffset = (short) (fragmentOffset & 0x1FFF);
        byte ttl = in.get();
        IpProtocol protocol = IpProtocol.fromType(in.get());
        in.getShort(); // Checksum
        Ipv4Address srcIp = Ipv4Address.decode(in);
        Ipv4Address dstIp = Ipv4Address.decode(in);
        in.reset();
        byte[] headerBytes = new byte[HEADER_LEN];
        in.get(headerBytes);
        if(ensureChecksum) {
            IpUtils.ensureInternetChecksum(headerBytes);
        } else {
            IpUtils.verifyInternetChecksum(headerBytes);
        }
        List<IpOption> options = new ArrayList<>();
        int optionsLength = (ihl - 5) * 4;
        while(optionsLength > 0) {
            IpOption option = IpOption.decode(in);
            options.add(option);
            optionsLength -= option.length();
        }
        if(optionsLength != 0) {
            throw new IllegalArgumentException("Could not decode options for the IPv4 header.");
        }
        return new Ipv4Header(dscp, ecn, len, identification, flags, fragmentOffset, ttl, protocol,
                srcIp, dstIp, options);
    }
}

File: src/main/kotlin/com/github/hirsivaja/ip/packet/header/Ipv4Header.kt
package com.github.hirsivaja.ip.packet.header

import com.github.hirsivaja.ip.DataArray
import com.github.hirsivaja.ip.IpUtils
import com.github.hirsivaja.ip.address.Ipv4Address
import com.github.hirsivaja.ip.protocol.IpProtocol
import java.nio.ByteBuffer

data class Ipv4Header(
    val ihl: UByte,
    val ecn: Byte,
    val dscp: Byte,
    val len: UShort,
    val identification: Short,
    val flags: Byte,
    val fragmentOffset: UShort,
    val ttl: UByte,
    val protocol: IpProtocol,
    val srcIp: Ipv4Address,
    val dstIp: Ipv4Address,
    val options: DataArray
) : IpHeader {

    constructor(
        ihl: UByte,
        ecn: Byte,
        dscp: Byte,
        len: UShort,
        identification: Short,
        flags: Byte,
        fragmentOffset: UShort,
        ttl: UByte,
        protocol: IpProtocol,
        srcIp: Ipv4Address,
        dstIp: Ipv4Address,
        options: ByteArray
    ) : this(ihl, ecn, dscp, len, identification, flags, fragmentOffset, ttl, protocol, srcIp, dstIp, DataArray(options))

    override fun encode(out: ByteBuffer) {
        out.mark()
        val versionIhl = (VERSION shl VERSION_SHIFT or ((options.length() / 4) + 5)).toByte()
        out.put(versionIhl)
        val dscpEcn = (ecn.toInt() and 0xFF or (dscp.toInt() shl DSCP_SHIFT)).toByte()
        out.put(dscpEcn)
        out.putShort(len.toShort())
        out.putShort(identification)
        val flagsFragmentOffset = (fragmentOffset.toInt() or (flags.toInt() shl FLAGS_SHIFT)).toShort()
        out.putShort(flagsFragmentOffset)
        out.put(ttl.toByte())
        out.put(protocol.type.toByte())
        val checksumPosition = out.position()
        out.putShort(0.toShort())
        srcIp.encode(out)
        dstIp.encode(out)
        out.put(options.data)
        val position = out.position()
        val headerBytes = ByteArray(HEADER_LEN)
        out.reset()[headerBytes]
        val checksum = IpUtils.calculateInternetChecksum(headerBytes)
        out.putShort(checksumPosition, checksum)
        out.position(position)
    }

    override fun generatePseudoHeader(): ByteArray {
        val out = ByteBuffer.allocate(PSEUDO_HEADER_LEN)
        srcIp.encode(out)
        dstIp.encode(out)
        out.put(0.toByte())
        out.put(protocol.type.toByte())
        out.putShort((len.toInt() - HEADER_LEN - options.length()).toShort())
        val outBytes = ByteArray(PSEUDO_HEADER_LEN)
        out.rewind()[outBytes]
        return outBytes
    }

    override fun length(): Int {
        return HEADER_LEN + options.length()
    }

    override fun pseudoHeaderLength(): Int {
        return PSEUDO_HEADER_LEN
    }

    override fun protocol(): IpProtocol {
        return protocol
    }

    companion object {
        fun decode(buf: ByteBuffer, ensureChecksum: Boolean) : Ipv4Header {
            buf.mark()
            val ihlVersion: Byte = buf.get()
            val ihl = (ihlVersion.toInt() and 0x0F).toUByte()
            val version = (ihlVersion.toInt() ushr VERSION_SHIFT).toByte()
            require(version.toInt() == VERSION) { "Unexpected version for IPv4 header! $version" }
            val ecnDscp: Byte = buf.get()
            val ecn: Byte = (ecnDscp.toInt() and 3).toByte()
            val dscp = (ecnDscp.toInt() ushr DSCP_SHIFT).toByte()
            val len: UShort = buf.getShort().toUShort()
            val identification: Short = buf.getShort()
            val flagsFragmentOffset: Short = buf.getShort()
            val flags: Byte = (flagsFragmentOffset.toInt() ushr FLAGS_SHIFT).toByte()
            val fragmentOffset: UShort = (flagsFragmentOffset.toInt() and 0x1FFF).toUShort()
            val ttl: UByte = buf.get().toUByte()
            val protocol: IpProtocol = IpProtocol.fromType(buf.get().toUByte())
            buf.getShort() // Checksum
            val srcIp: Ipv4Address = Ipv4Address.decode(buf)
            val dstIp: Ipv4Address = Ipv4Address.decode(buf)
            buf.reset()
            val headerBytes = ByteArray(HEADER_LEN)
            buf[headerBytes]
            if(ensureChecksum) {
                IpUtils.ensureInternetChecksum(headerBytes)
            } else {
                IpUtils.verifyInternetChecksum(headerBytes)
            }
            val options = ByteArray((ihl.toInt() - 5) * 4)
            buf[options]
            return Ipv4Header(ihl, ecn, dscp, len, identification, flags, fragmentOffset, ttl, protocol, srcIp, dstIp, options)
        }

        const val VERSION: Int = 4
        const val HEADER_LEN: Int = 20
        const val PSEUDO_HEADER_LEN: Int = 12
        const val VERSION_SHIFT: Int = 4
        private const val DSCP_SHIFT: Int = 2
        private const val FLAGS_SHIFT: Int = 13
    }
}
